core/workspace_svg.js

```diff
@@ -449,6 +458,25 @@ Blockly.WorkspaceSvg.prototype.setMarkerSvg = function(markerSvg) {
   this.markerManager_.setMarkerSvg(markerSvg);
 };
 
+/**
+ * Add a plugin to the list of plugins for this workspace. If it is of certain
+ * types then add it to all lists that it applies to.
+ * @param {string} type The type of the plugin.
+ * @param {Object} pluginClass The constructor for a plugin.
+ * @param {Object} params The other params for the plugin.
+ */
+Blockly.WorkspaceSvg.prototype.addPlugin = function(type, pluginClass, params) {
+  if (!this.plugins[type]) {
+    this.plugins[type] = [];
+  }
+  var newPlugin = new pluginClass(this, params);
+  newPlugin.init();
+  this.plugins[type].push(newPlugin);
+
+  // List of all plugins for dispose.
+  this.plugins['all'].push(newPlugin);
```

Comment:
Why is this needed, wouldn't it be enough to go through `Object.keys` on `this.plugins` to get all of them? or am i missing something?

Reply:
Yeah I think I was trying to communicate something else and failed lol 
I was trying to get at that plugins can have some kind of hierarchy:
plugin ->  uiPlugin -> some_other_plugin
If we have some_other_plugin then we want it to go in it's own list as well as the uiPlugin list so that we can easily get all the things that need to change without having to search through.

Reply:
I see, so in that case you would register a toolbox as both: 
``workspace.addPlugin('toolbox', Blockly.Toolbox)``
and 
``workspace.addPlugin('uiplugin', Blockly.Toolbox)``

or perhaps a variation of the above with an array: 
``workspace.addPlugin(['toolbox', 'uiplugin'], Blockly.Toolbox)``
?

Reply:
Yeah basically. I kind of wanted to take away that work from the developer and do it for them in the cases that we know. For example, toolbox is always going to be a uiplugin and we can add it for them. But this might be too convoluted on our end.